### DISCUSSION TITLE
vNext - Features and ideas

### DISCUSSION THREAD

**Discussion by enkelmedia** [enkelmedia/TheDashboard#106] [General]
After some proof of concept work during CodeCabin 2024 I showed a more modular version of The Dashboard. The idea is transmission into a "dashboard facilitator" with some simple "out of the box" widget to show but to allow for other packages and site implementors to extend the package.

This development effort will target the new backoffice aka Bellissima.

Here is a demo of the flexible "widget-based" UI:

![The dashboard vNext](https://raw.githubusercontent.com/enkelmedia/TheDashboard/refs/heads/v14/dev-new/documentation/dashboard-vnext-wip.gif)

Here are some of the ideas/features that have been raised:
* The new dashboard is built up using independent "widgets"
* The package should ship with a limited (and manageable) set of default widgets, but allow for any package developer or site implementor to extend with new types of widgets 
* Each widget should support "settings" (could be filters etc)
* Some sort of mechanism for "global settings" should be in place so that a widget could contain global information that individual users cannot change. E.g. a "Support contact" widget that is managed by administrators.
* The dashboard configuration (placement, size, etc) should be stored on a User-level
* Site implementors should be able to provide a "default" configuration that will be used until the individual user make any adjustments.
* We need a custom table (or some other mechanism) to store the user-level configuration.

I want to start by collecting ideas and feedback and, at the same time also try to keep the scope in a manageable size. The initial scope will not cover any new type of widgets. I'm looking for ideas/feedback on the setup and over all structure.

Sketched concept by Markus Johansson and Joe Glombek from CodeCabin 2024:
![image](https://github.com/user-attachments/assets/b0606935-e031-428c-ba7e-adbe22adf5cd)

**Comment by Rockerby** [comment]
I think this sounds like a great idea! Making the "widgets" composable to allow people to extend it would be a great addition.

Bluesky thinking would be to then allow users to drag, resize and reorganise the dashboard to their liking with a library of components to add/remove. Per user!

How can I help?

Leaving this here so I don't forget I found it - https://gridstackjs.com/

**Comment by LottePitcher** [comment]
I loved this idea at Codecabin, and still do!

I see "The package should ship with a limited (and manageable) set of default widgets, but allow for any package developer or site implementor to extend with new types of widgets"

I understand you don't want to become responsible for many community widgets but perhaps there could be a separate community managed repository of dashboard widgets. A bit like the TagHelpers project that is hosted on the Umbraco Community org: https://github.com/umbraco-community/Our-Umbraco-TagHelpers.  Might make sense over having lots and lots of separate packages for each widget people want to share. 

I appreciate I'm getting ahead of myself here, just all excited again! 

If there's any help that either the Package Team or DevRel can offer with this then please let me know!

**Comment by enkelmedia** [comment]
Lovely to see the engagement! :) 

I did start some work on this with GridStackJS in a branch in the repo, but I also realized that the foundation for this would make a lot of sense to put into the core CMS - especially if a package developer would like to add a “widget”, they wouldn’t have to create a separate package to target “The Dashboard” to provide a widget.

I’ve been talking with Mads Rasmussen about the idea of having the widgets as a extension type and he has been prototyping a new extension type with the working name “dashboardApp”, there is a branch with some initial work here:

https://github.com/umbraco/Umbraco-CMS/tree/v16/poc/dashboard-apps

The idea is that the CMS would contain the “grid” for the widgets, possibly based on code similar to the code that is powering the block grid (for resize etc).

When we talked during CodeGarden, one idea was that the core CMS would only contain this foundation while a package (like e.g The Dashboard) could provide any number of “apps”.

I would love to see this in the core and for this package to transition into being the “go to” for these standard widgets. I also agree with @LottePitcher that it would be cool to “coordinate” widgets from the community in some way.

I’ve offered Mads to help out but I think it needs a little more “direction”. I guess that if DevRel could raise this internally it might help to get it on the roadmap. 

It would also be helpful if we could get some validation/feedback on the thoughts outlined in the WIP PR that Mads created during CG: 

https://github.com/umbraco/Umbraco-CMS/pull/19575

**Comment by enkelmedia** [comment]
@Rockerby, @LottePitcher 

Would love to hear if you have any input on the progress going on here:

https://github.com/umbraco/Umbraco-CMS/pull/19575
